ARCHITECTURE.md

```diff
@@ -188,3 +194,34 @@ kURL automation, more specifically the `import-external-addons` GitHub action, p
 New versions are published to the external add-on registry and packages are copied from the source and stored in the kURL S3 bucket.
 
 The kURL API merges the external add-on registry with its internal list of add-on versions, making them available to the end-user.
+
+## Deployment and Releases
+
+Upon releasing kURL, scripts and add-on package archives are built and uploaded to to the kURL object storage bucket along with metadata including the Git sha from which they were generated.
+Once complete, the VERSION file is updated to point to the current version of kURL.
+The API makes use of this VERSION file to resolve the scripts and add-on packages.
+
+### Production Workflow
+
+Production release are triggered by running the command `make tag-and-release` or pushing a tag in the format "v*.*.*".
+Production releases are uploaded to the object storage bucket at prefix `dist`, for example https://kurl-sh.s3.amazonaws.com/dist/v2022.04.19-0/.
+Before building add-on packages, the workflow will first check if there were any changes made to the add-on since the previous production release based on metadata included with the add-on package.
+If no changes were made, the package will be copied from the previous production release to optimize for build times.
+Next, the workflow will check if there were any changes made since the previous staging release and copy from staging if no changes were made.
+The is to account for a scenario where a commit were being tagged other than what is at the HEAD of main.
+If changes were made to both production and staging, the package will be built from source and uploaded.
+Finally, VERSION file is updated at https://kurl-sh.s3.amazonaws.com/dist/VERSION to point to the new production version.
+Historical production releases are never removed from the object storage bucket.
+
+### Staging Workflow
+
+Staging releases are triggered on merge to main.
+Staging release versions use the most current release version tag and append prerelease `-dirty`, for example v2022.04.19-0-dirty.
+Due to this versioning scheme, staging releases will overwrite the previous staging release if no production release occurred prior.
+This is intentional to optimize for storage costs.
+Staging releases are uploaded to the object storage bucket at prefix `staging`, for example https://kurl-sh.s3.amazonaws.com/staging/v2022.04.19-0-dirty/.
+Before building add-on packages, the workflow will first check if there were any changes made to the add-on since the previous staging release based on metadata included with the add-on package.
+If no changes were made, the package will be copied from the previous staging release to optimize for build times.
+If changes were made, the package will be built from source and uploaded.
+Finally, VERSION file is updated at https://kurl-sh.s3.amazonaws.com/staging/VERSION to point to the new staging version.
+Historical staging releases are never removed from the object storage bucket.
```

Comment:
Have we a reason to keep all artefacts for the staging ones? Have we any scenario where would be required to use the old ones? If not, could we not remove to reduce the cost to keep all?

Reply:
Answered in the call

Reply:
No. I propose we change this to delete staging releases after some measure